### ISSUE TITLE
Pose priors in BA cost function

### ISSUE THREAD

**Issue by victor-amblard** [openMVG/openMVG#1694]
Hi Pierre,

I delved into the code to try to understand the use of motion priors with `IncrementalSfM2` and I have some doubts about what the `-P` option actually does. It seems that when using EXIF data and following the instructions from https://openmvg.readthedocs.io/en/latest/software/Geodesy/geodesy/ everything works but if you try to manually add poses as part of the json file in the `views` section with `center`, `center_weight` and `use_pose_center_prior` fields (for example IMU data), the priors don't seem to be really taken into account completely. Actually from what I saw the `openMVG::sfm::ViewPriors::load` which loads these parameters is never called by `IncrementalSfM2` and more generally those parameters are never read.

To be more specific, in the file `sfm_data_BA_ceres.cpp` I woud like `b_usable_prior` to be `true` so that I can add my pose priors in the cost function _via_ `PoseCenterConstraintCostFunction` struct. But in order for it to be `true` I need to have `sfm_data.views` of type `ViewPriors` (which inherits from `View`) but it seems to be initialized with `View`...

So basically my question is the following: is there a way to make views of type `ViewPriors` instead of `View` when `-P` option is enabled?

Thanks!

**Comment by pmoulon** [comment]
When you use -P on SfmInitImageListing you will have the View of type ViewPriors and then you can run SfM and the camera center will be constrained

**Comment by pmoulon** [comment]
See here https://github.com/openMVG/openMVG/blob/develop/src/software/SfM/main_SfMInit_ImageListing.cpp#L409

**Comment by victor-amblard** [comment]
Thanks for your answer. In order to enter this `if`you need EXIF data because you need gps_info to be `true`...

**Comment by pmoulon** [comment]
Ok so I would create a fake scene with GPS prior (adding fake exif data to your image) and then adding the true one and run the pipeline.
Sfm_InitImageListing is just a way to create the initial JSON file that describes your scene, you could also use Python to create this sfm_data.json file

**Comment by victor-amblard** [comment]
Thanks for the tip, haven't thought of that. 
I ended up adding an attribute `viewP` of type `ViewsPriors` to `sfm_data` struct and using `openMVG::sfm::ViewPriors::load` slightly modified called by `archive(cereal::make_nvp("views", data.viewsP));` in addition to the `archive(cereal::make_nvp("views", data.views));`  (in `sfm_data_io_cereal.cpp` so I directly load priors from the json file this way.

**Comment by pmoulon** [comment]
Glad you find a way
